### ISSUE TITLE
All files downloaded on every build even if nothing has changed

### ISSUE THREAD

**Issue by dottodot** [NekR/offline-plugin#168]
A bit new to service workers and trying to get my head around it, but when I build my project a second time with no changes I'm finding the the service worker is updated somehow and when I reload the site the process of installing all the assets starts again.

I thought that if there were no changes there should not be any need for the service worker to update or have to re-download all the assets again.

I've tried various combinations of settings and they all seem to behave the same, but here's my current settings.

```
        new OfflinePlugin({
            caches: {
                main: ['js/*.js', 'css/*.css'],
                additional: [':rest:', ':externals:'],
                optional: []
            },
            version: pjson.version,
            excludes: ['**/.*', '**/*.map', '../views/*.handlebars', 'fonts/**', 'img/flags/**'],
            AppCache: false,
            ServiceWorker: {
                events: true,
                output:'service-worker.js'
            }
        })
```

**Comment by NekR** [comment]
It should download only that what was changed. How do you determine if it downloads files or not?

**Comment by dottodot** [comment]
In the network activity you can see when the assets are downloaded and then cached. This only happens when the service worker is installed and in my case every time I do a new build, dispite there not being any changes.

**Comment by GGAlanSmithee** [comment]
@dottodot you should probably share your whole webpack config. For example; do you use hashed chunks? If so, maybe the hash changes between builds for whatever reason?

**Comment by NekR** [comment]
@GGAlanSmithee `offline-plugin` generates its own hashes in since 4.5 and doesn't rely on webpack's generated hashes. Though, if at least any byte on a bundle changes -- you get a new hash for it.

@dottodot what you can do here is do a built and then open `service-worker.js` and copy value of `__wpo` variable. Then do another build and copy it again. Then share it here so we could compare hashes or do compare them yourself.

Basically, if hashes change then file is redownloaded. If hashes do not change and files are redownloaded then that's a bug.

As @GGAlanSmithee said, it probably makes sense to share your webpack config for easier debug, e.g. if you bundle's name changes between each built. That shouldn't affect how `offline-plugin` caches things, but it would be good to have such info here.

So, please provide (if possible):
* Your webpack config (you may send it my email if you don't feel sharing it publicly)
* `__wpo` variable value from `service-worker.js` for a build 1
* `__wpo` variable value from `service-worker.js` for a build 2 (the next)

**Comment by GGAlanSmithee** [comment]
@NekR ah okay, wasn't aware, sorry for the noise. (Just too be sure, can you also confirm that you are using version >= 4.5 @dottodot)

**Comment by dottodot** [comment]
Yes I'm using 4.5 and have just compared between 2 builds and can see there are differences caused by CopyWebpackPlugin and the image files that are added to the service worker aren't always added in the same order, so not sure if that's the cause.

Anyway here my config and the __wpo info
Config
https://gist.github.com/dottodot/298f8152d7c01f3739ca44018a0952f4
Build 1
https://gist.github.com/dottodot/261012df25182cb7c3912acf57c12f01
Build 2
https://gist.github.com/dottodot/81af2c93dc1c41311ae4e286b5ba12f1

**Comment by NekR** [comment]
@GGAlanSmithee no problem :-)
@dottodot order doesn't actually matter. I'll try to debug it with your config. Just to clarify other things: Do you see it in production? If in development, when on which server? webpack-dev-server?

**Comment by dottodot** [comment]
Happens on production as well and not using webpack-dev-server at the moment.

**Comment by dottodot** [comment]
Think it's definitely somehow related to CopyWebpackPlugin as when I do a build that just has js and css files it works as expected.

**Comment by NekR** [comment]
Hmm.. That's great info actually, thanks for providing it. Unfortunately plugins incompatibility happens all the time, but I'll to figure that out.

**Comment by haf** [comment]
I think I may be experiencing this issue.

 1. Visit staging-001.qvitoo.com
 2. Look at the service worker doing its thing
 3. After we deploy (which we do once in a while), the first load after a deploy crashes, because the service worker re-downloads scripts that are not there any longer. `index.html` is `Cache-Control: no-cache`; so the index file points at valid JS files.
 4. Pressing CMD+R; the old SW cache is still in effect but makes the same network request despite having contents cached, thus crashing. (see image below)
 5. Pressing SHIFT+CMD+R; everything is purged from the caches, the site loads fine.

<img width="1275" alt="screen shot 2017-01-30 at 09 58 09" src="https://cloud.githubusercontent.com/assets/193115/22417225/8b724a60-e6d3-11e6-95b8-9bb59fa27c72.png">

For all requests that are non-matches to files or API functions, we just serve the index file, which is why the error is about syntax in the JS file; but the actual error is something like the SW loading old stuff over the network without consideration for the new hashes.

**Comment by NekR** [comment]
Hey @haf, I don't think your issue is related to this issue. First of all, I don't even see your website registering a SW. It has `https://staging-001.qvitoo.com/sw.js` file, doesn't register it. But I see in the console that it's being registered, so I assume you just disabled it for now.

> After we deploy (which we do once in a while, have a look at /health to see the version), the first load after a deploy crashes, because the service worker re-downloads scripts that are not there any longer. index.html is Cache-Control: no-cache; so the index file points at valid JS files.

First load after a deploy may crash only if requests are going directly to the servers, i.e. to the network. With ServiceWorker they are cached and request go to the cache. In your case chunks are requested from the network (console shows it) but `index.html` seems to be loaded from the SW cache. So, it loads old `index.html` which points to old scripts which doesn't exist anymore, so server returns HTML and page doesn't work.

It seems to not be related to ServiceWorker updates because they don't work as your described. When you first visit the page after new deploy -- you get an old page loaded from the cache. The new page and its assets are being downloaded in background and that new page will be available for usage only when **all tabs of the website will be closed**. Once all tabs closed -- new page takes place and caches of old page are removed. Here are more info about SW update process: https://github.com/NekR/offline-plugin/blob/master/docs/updates.md

`Cmd+R` doesn't close the tab, hence you don't get the new cache. `Cmd+Shift+R` doesn't purge cache, it bypassed ServiceWorker for this particular tab for this particular refresh. Once you refresh again with just `Cmd+R` you will have the same problem.

Also, since you are using unique file names for each build, it's a good idea to keep your old files on the server and don't delete them. Even though you server `index.html` without caching, it still somehow may ended up cached or people will be navigating your website without a reload and will get broken routes because old JS files aren't in place.

**Comment by NekR** [comment]
@haf also, you are using an old version of `offline-plugin`, presumably `3.*` which has `updateStrategy: 'all'` by default instead of `'changed'` which was in `4.*`. Please update :-)

**Comment by haf** [comment]
> First load after a deploy may crash only if requests are going directly to the servers, i.e. to the network. With ServiceWorker they are cached and request go to the cache. In your case chunks are requested from the network (console shows it) but index.html seems to be loaded from the SW cache.

Why are the requests going to the network and not the cache then? It would seem that they should go to either-or.

**Comment by NekR** [comment]
As I said, if requests are going to network only then those URLs aren't cached. Also, I you think there are problems there with the plugin, then please open a separate issue since it isn't related to copy-webpack-plugin issue (this one). Thanks.

**Comment by haf** [comment]
@NekR Thanks. I created a new issue.
